File: dynamic_listings/C++/default/design_patterns/strategy.cpp
class Strategy{
    // This class defines the strategy interface the client will refer to

    public:
        // This algorithm will be implemented by the subclasses
        virtual void algorithm() = 0;
};

class ConcreteStrategy1: public Strategy{
    public:
        void algorithm() override{
            // Real implementation of the algorithm
            // DO STUFF
        }
};

class ConcreteStrategy2: public Strategy{
    public:
        void algorithm() override{
            // Real implementation of the algorithm
            // DO STUFF SLIGHTLY DIFFERENTLY
        }
};

// Example Usage
int main(){
    Strategy* to_execute;
    if (condition){
        to_execute = new ConcreteStrategy1();
    }else{
        to_execute = new ConcreteStrategy2();
    }
    to_execute->algorithm();  // This will execute 1 or 2 depending on "condition"
}

File: dynamic_listings/rust/default/design_patterns/strategy.rs
trait Strategy {
    // This class defines the strategy interface the client will refer to

    // This algorithm will be implemented by the subclasses
    fn algorithm(&self);
}

struct ConcreteStrategy1;

impl Strategy for ConcreteStrategy1 {
    fn algorithm(&self) {
        // Real implementation of the algorithm
        // DO STUFF
    }
}

struct ConcreteStrategy2;

impl Strategy for ConcreteStrategy2 {
    fn algorithm(&self) {
        // Real implementation of the algorithm
        // DO STUFF SLIGHTLY DIFFERENTLY
    }
}

// Example Usage
fn main() {
#   let condition = true;
    let to_execute = if condition {
        Box::new(&ConcreteStrategy1 as &dyn Strategy)
    } else {
        Box::new(&ConcreteStrategy2 as &dyn Strategy)
    };
    to_execute.algorithm(); // This will execute 1 or 2 depending on "condition"
}
